src/hal/src/command/mod.rs

```diff
@@ -15,53 +15,95 @@ pub use self::raw::RawCommandBuffer;
 pub use self::renderpass::*;
 pub use self::transfer::*;
 
-/// Thread-safe finished command buffer for submission.
-pub struct Submit<B: Backend, C>(pub(crate) B::CommandBuffer, PhantomData<C>);
-unsafe impl<B: Backend, C> Send for Submit<B, C> {}
-
-impl<B: Backend, C> Submit<B, C> {
+/// Trait used to represent a single submission that can be added to a `Submission`.
+pub unsafe trait Submit<B: Backend, C>: Send {
```

Comment:
There is no point to have a trait with methods if all the implementations do them in the same way. How about:
```rust
trait Shot {}
enum OneShot {}
enum MultiShot {}
impl Shot for OneShot {}
impl Shot for MultiShot {}

pub struct Submit<B: Backend, C, S>(pub(crate) B::CommandBuffer, PhantomData<(C, S)>);
impl<B: Backend, C, S: Shot> Submit<B, C> {
  ...
}
```

Reply:
Sure, that works. I still think `Shot` is kinda a strange name though...